src/Symfony/Bridge/Twig/Node/SearchAndRenderBlockNode.php

```diff
@@ -24,88 +28,92 @@ class SearchAndRenderBlockNode extends FunctionExpression
-                    // All other functions than "label" expect the variables
-                    // in the second argument
-                    $label = null;
-                    $variables = $arguments[1];
-                    $labelIsExpression = false;
-                }
-
-                if (null !== $variables || $labelIsExpression) {
-                    $compiler->raw(', ');
 
-                    if (null !== $variables) {
-                        $compiler->subcompile($variables);
+                        // Insert the label argument into the array
+                        $variables->addElement($label, $labelKey);
+                        $arguments['variables'] = $variables;
                     }
-
-                    if ($labelIsExpression) {
-                        if (null !== $variables) {
-                            $compiler->raw(' + ');
-                        }
-
-                        // Check at runtime whether the label is empty.
-                        // If not, add it to the array at runtime.
-                        $compiler->raw('(twig_test_empty($_label_ = ');
-                        $compiler->subcompile($label);
-                        $compiler->raw(') ? array() : array("label" => $_label_))');
+                } else {
+                    // The label argument is not a constant, but some kind of
+                    // expression. This expression needs to be evaluated at runtime.
+                    // Depending on the result (whether it is null or not), the
+                    // label in the arguments should take precedence over the label
+                    // in the attributes or not.
+
+                    // Check at runtime whether the label is empty.
+                    // If not, add it to the array at runtime.
+                    $labelExpr = new ArrayExpression(array(new ConstantExpression('label', $lineno), $label), $lineno);
+                    $condition = new TestExpression($label, 'empty', new Node(), $lineno);
+                    $conditional = new ConditionalExpression($condition, new ArrayExpression(array(), $lineno), $labelExpr, $lineno);
+
+                    if (count($arguments['variables'])) {
+                        $arguments['variables'] = new FilterExpression(
+                            new Node(array($conditional)),
+                            new ConstantExpression('merge', $lineno),
+                            $arguments['variables'],
+                            $lineno
+                        );
+                    } else {
+                        $arguments['variables'] = $conditional;
                     }
                 }
+
+                unset($arguments[1]);
+            }
+        } else {
+            if (isset($arguments[1])) {
+                $arguments['variables'] = $arguments[1];
+                unset($arguments[1]);
             }
         }
 
-        $compiler->raw(')');
+        $this->setNode('arguments', new Node(array_merge(
```

Comment:
to avoid weird side effects, the state of the node should be reset after calling the parent logic. Otherwise, compiling the same node object multiple time might break